### PR TITLE
Avoid unnecessary interlocked in ManualResetValueTaskSourceCore

### PR DESCRIPTION
**Pull request by stephentoub**
When a ManualResetValueTaskSourceCore is used as the implementation for a ValueTask, it's set up for the operation, and then two things happen: a callback is hooked up at some point, and the operation completes at some point.  The former generally occurs before the latter, however there is a race condition, and so both paths currently use an Interlocked.CompareExchange to coordinate.  But that means that we always end up with two CompareExchange operations.  In the common path, there's no contention between these, and so we can avoid one of the CompareExchanges by first doing a normal read of the target field (we were already doing that read in one of the two cases, but we weren't taking advantage of it).

cc: @kouvel, @tarekgh

### CODE REVIEW

**Comment by tarekgh** [comment]
>         public void OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)

To make sure I understand whole things here, is it possible OnCompleted will get twice in 2 different threads with different continuation actions? and so, would that be permissible? reading the code it is not permissible but want to know why?

---
Refers to: src/System.Private.CoreLib/shared/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs:108 in 9f456be. [](commit_id = 9f456be16b87347e8c540cdb141a634b5654f1b0, deletion_comment = False)

**Comment by stephentoub** [comment]
> To make sure I understand whole things here, is it possible OnCompleted will get twice in 2 different threads with different continuation actions?

With correct usage, no.

> and so, would that be permissible?

No.

>  reading the code it is not permissible but want to know why?

Part of the design of `ValueTask`/`ValueTask<TResult>` is that they don't support multiple awaiters, whether serialized or concurrent, and that's the only way you could end up with multiple OnCompleted calls.  They don't support them because we need to be able to support ValueTasks backed by IValueTaskSources that become reusable for a different operation once the await on the ValueTask completes.

**Comment by stephentoub** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests please
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests please
@dotnet-bot test Windows_NT x64 Release CoreFX Tests please

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs [hunk 1]
```diff
@@ -142,17 +142,22 @@ public void OnCompleted(Action<object> continuation, object state, short token,
             // awaited twice concurrently), _continuationState might get erroneously overwritten.
             // To minimize the chances of that, we check preemptively whether _continuation
             // is already set to something other than the completion sentinel.
-            object currentContinuation = _continuation;
-            if (currentContinuation != null &&
-                !ReferenceEquals(currentContinuation, ManualResetValueTaskSourceCoreShared.s_sentinel))
+
+            object oldContinuation = _continuation;
+            if (oldContinuation == null)
```
[the remark this answers is not in the document]
Comment (kouvel) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs]: Nevermind, it looks like it's doing the same thing as before

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs]: No.  If we read a non-null value, that means it's the completion sentinel, which means the operation has already completed by the time we awaited, in which case the subsequent code will just queue the supplied continuation/state, ignoring _continuation.  When the object is then reused, the user will call Reset on it, and _continuation will be set to null in preparation for the next operation.

[the remark this answers is not in the document]
Comment (kouvel) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs]: I see, thanks

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs]: > Nevermind, it looks like it's doing the same thing as before

Yes, with the exception that we skip the Interlocked.CompareExchange if we know it's going to fail anyway, and in that case we also avoid writing to the state field, as it's also going to be ignored.

**Comment by stephentoub** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests please
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests please
@dotnet-bot test Windows_NT x64 Release CoreFX Tests please

**Pull request merged**